gdk4/src/texture.rs

```diff
@@ -18,3 +21,51 @@ impl<O: IsA<Texture>> TextureExtManual for O {
         }
     }
 }
+
+impl Texture {
+    pub fn for_pixbuf_future(
+        pixbuf: &'static gdk_pixbuf::Pixbuf,
+    ) -> Pin<Box<dyn Future<Output = Option<Texture>> + 'static>> {
+        Box::pin(async move {
+            let (sender, receiver) = futures_channel::oneshot::channel();
+
+            let sender = std::cell::Cell::new(Some(sender));
+
+            let task = gio::Task::new(
+                Some(&pixbuf.upcast_ref()),
+                gio::Cancellable::NONE,
+                move |result, _| {
+                    let texture = result
+                        .downcast_ref::<gio::Task>()
+                        .unwrap()
+                        .propagate_value()
+                        .unwrap()
+                        .get::<Texture>()
+                        .unwrap();
+                    if let Some(m) = sender.replace(None) {
+                        let _result = m.send(texture);
+                    }
+                },
+            );
+
+            let task_func = move |task: &gio::Task,
+                                  source_object: Option<&glib::Object>,
+                                  _: Option<&gio::Cancellable>| {
+                let pixbuf = source_object
+                    .unwrap()
+                    .downcast_ref::<gdk_pixbuf::Pixbuf>()
+                    .unwrap();
+                let texture = Texture::for_pixbuf(&pixbuf);
+                task.return_value(&texture.to_value());
```

Comment:
@pbor The API is unsound, oh no. You can pass a non-`Send` value from the `run_in_thread()` callback back to the main thread.

Reply:
:(